### ISSUE TITLE
Cache-aligned atomic var 

### ISSUE THREAD

**Issue by bartoszmodelski** [ocaml-multicore/saturn#19]
When writing lock-free code, It's quite easy to lose performance due to false sharing. It happens when multiple variables occupy the same cache line and atomic operations acting on them cannot be executed in parallel. 

How much of an impact can it actually have? @anmolsahoo25 found a 22% improvement in spsc queue with indices aligned to cache lines. With just 2 threads! 

However, we don't have a good story for preventing false sharing. OCaml doesn't allow forcing memory layout. We could add unused fields to the struct as padding but this feels quite fragile, e.g. optimisation could fall apart with compiler update. But it can be done with C bindings (see Anmol's experiment https://github.com/anmolsahoo25/aligned_int). 

It would be great to build out something similar in the long-term to stop leaving performance on the table.

**Comment by kayceesrk** [comment]
Good point. I'm in favour of having a cache aligned atomic var in the library. 

What API do you have in mind? It may be useful to see whether we could do a `make_cache_aligned_atomic : 'a -> 'a Atomic.t` and hence reuse the rest of the functions from the Atomic module.

**Comment by bartoszmodelski** [comment]
For my uses `AlignedInt` would be good enough. I agree there's value in a more general interface but I don't really have a feel for how hard it'd be. Wouldn't mind giving it a go later on though.

**Comment by kayceesrk** [comment]
IIRC Anmol's prototype allocated memory for the aligned atomic int off heap. Hence, they'd have to be manually freed?

**Comment by bartoszmodelski** [comment]
Yes, it uses `posix_memalign` to get memory. But I'd _guess_ we could hack something similar on heap (for all non-pointers) by allocating a contiguous block of memory of size 2*cache_line and using some region in the middle of it as atomic value. 

FWIW I checked how glibc's posix_memalign works. It allocates a larger chunk with malloc (one that's guaranteed to have an aligned value in it) and then frees the leading and trailing space.

**Comment by kayceesrk** [comment]
> Yes, it uses posix_memalign to get memory. But I'd guess we could hack something similar on heap (for all non-pointers) by allocating a contiguous block of memory of size 2*cache_line and using some region in the middle of it as atomic value.

This is what I had in mind.

**Comment by lyrm** [comment]
I have made some very basic tests to measure false sharing. I tried : 

- Test 1 (with false sharing) : multiple domains incrementing their own cell in an array but without any kind of padding,
- Test 2 (avoiding false sharing with padding) : same but with some empty cells between the ones with a counter in it,
- Test 3 (avoiding false sharing with reference) :  having an array of `int ref` instead of just `int`. This way the cells are read only.

What I get is a difference of around ~20% of runtime between the two first tests (with 2 domains), and around the same performance between test 2 and  test 3. 

I pushed the tests I did [here](https://github.com/lyrm/testing_lf_ideas) if you want to have a look. 

Note : these are very basic tests. In particular, I am not trying to avoid any GC related run time variance except by printing a median and an average runtime value.

**Comment by bartoszmodelski** [comment]
I tried the idea with allocating a wider block and selecting atomic out of there. See the code below. It allocates a block of 3 cache lines, finds an aligned spot inside, and chops the allocation into 3 blocks of required sizes. 

```
value caml_make_aligned(value val) 
{
  size_t size = cache_line_size();

  /* get 3x cache line */
  value chunk = caml_alloc_shr_check_gc(3 * size / 8 - 1, 0); 

  for (int i = 0; i < 3 * size / 8 - 1; i++) {
    ((uint64_t*)chunk)[i] = Val_int(0);
  }

  size_t overshot = chunk % size;
    
  value var1 = chunk; 
  value var2 = var1 + 2 * size - overshot + 8; /* +8 for header */
  value var3 = var2 + 16; 

  size_t size_var1 = (var2 - 8) - var1; /* -8 for header */
  size_t size_var2 = (var3 - 8) - var2;
  size_t size_var3 = ((var1 + 3 * size - 8) - var3);

  *Hp_val(var1) =  Make_header(size_var1 / 8, 0, 0);
  *Hp_val(var2) =  Make_header(size_var2 / 8, 0, 0);
  *Hp_val(var3) =  Make_header(size_var3 / 8, 0, 0);

  caml_initialize(&Field(var2, 0), val);

  return var2;
}
```
## Result 

* Variables are aligned and stay that way (_shr_ sends them straight to major). 

* They are getting collected as finalizers on _var1, var3_ are getting called. _live_blocks_ looks good after collection. 

BUT other garbage collector stats are mangled. Number of live words seems far too high (40% of _major_words_), number of fragments goes negative. See example below. The test allocated 100k aligned atomics.

```
minor_words:    1384
promoted_words:  117
major_words:    2500184

top_heap_words: 2418339
heap_words:       36864
live_words:     1100819
free_words:       36612
largest_free:         0
fragments:      -1100567

live_blocks: 33
free_blocks: 0
heap_chunks: 0
```

I haven't dug much deeper into it but on a glance I'm not that surprised that multiplying blocks breaks things. I'd be very happy to hear that my C code is broken, but if that's not the case and we would have to do a compiler patch, the following section describes a nicer solution. 

## Alternative solution? 

Whenever we allocate a block with `alloc_shr`, it calls `caml_shared_try_alloc` and there, depending on the size of the allocation we're gonna either do `pool_allocate` or [`large_allocate`](https://github.com/ocaml/ocaml/blob/97eb1bc9b9cf9a2b1ae14f8b8b5d42b27c1a8eac/runtime/shared_heap.c#L378). `large_allocate` just calls `malloc` to get new memory and does some accounting for the allocation to be freed eventually. 

We could add a flag to just call [`aligned_alloc`](https://en.cppreference.com/w/c/memory/aligned_alloc) instead! It should work out of the box, it would be more future-proof (compactions), other people could use it, not sure how receptive upstream would be to that.

**Comment by bartoszmodelski** [comment]
Another data-point from @sadiqj: the Java-way seems to be just do the padding trick ([example](https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/MpscCompoundQueue.java#L30)). 

I had a look at Go, and it's pretty much the same, but with a special type for doing the padding ([CacheLinePad](https://go.dev/src/internal/cpu/cpu.go)). I like it, as it's now on the compiler to provide cache line size for every architecture and there's a way to tell compiler to not reorder the struct. 

cc @polytypic, who might be interested in this issue

**Comment by sadiqj** [comment]
We could have an off-heap pointer to some aligned memory? As long as the block has a valid header then it's fine. We'd just need to take care of free'ing it eventually.

Alternatively, we could also add support for not boxing Atomics and have some atomic operations on `Obj` which means padding a record will work fine.

**Comment by kayceesrk** [comment]
> We could have an off-heap pointer to some aligned memory? As long as the block has a valid header then it's fine. We'd just need to take care of free'ing it eventually.

If the data in the atomic object is a primitive value, then the block could be tagged as `Abstract_tag`. GC will not trace the object.

It would have been nice if OCaml custom blocks (https://v2.ocaml.org/manual/intfc.html#s%3Ac-custom) included an operation to tell the GC how to trace the custom object to find pointers.

**Comment by Sudha247** [comment]
Cache-aligned atomic was added to OCaml in https://github.com/ocaml/ocaml/pull/12212 and will be released in OCaml 5.2. Closing this issue.
